pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,329 @@
+	"k8s.io/kubernetes/pkg/util"
+	utilErrors "k8s.io/kubernetes/pkg/util/errors"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+
+	"github.com/golang/glog"
+)
+
+// ProbeVolumePlugins is the entry point for plugin detection in a package.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&downwardAPIPlugin{}}
+}
+
+const (
+	downwardAPIPluginName = "kubernetes.io/downwardapi"
+)
+
+// downwardAPIPlugin implements the VolumePlugin interface.
+type downwardAPIPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &downwardAPIPlugin{}
+
+func (plugin *downwardAPIPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *downwardAPIPlugin) Name() string {
+	return downwardAPIPluginName
+}
+
+func (plugin *downwardAPIPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.VolumeSource.DownwardAPI != nil
+}
+
+func (plugin *downwardAPIPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, pod, opts, mounter)
+}
+
+func (plugin *downwardAPIPlugin) newBuilderInternal(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	v := &downwardAPIVolume{volName: spec.Name,
+		pod:     pod,
+		plugin:  plugin,
+		opts:    &opts,
+		mounter: mounter}
+	v.fieldReferenceFileNames = make(map[string]string)
+	for _, fileInfo := range spec.VolumeSource.DownwardAPI.Items {
+		v.fieldReferenceFileNames[fileInfo.FieldRef.FieldPath] = fileInfo.Name
+	}
+	return v, nil
+}
+
+func (plugin *downwardAPIPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, mounter)
+}
+
+func (plugin *downwardAPIPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &downwardAPIVolume{volName: volName,
+		pod:     &api.Pod{ObjectMeta: api.ObjectMeta{UID: podUID}},
+		plugin:  plugin,
+		mounter: mounter}, nil
+}
+
+// downwardAPIVolume handles retrieving metadata from the API server
+// and placing them into the volume on the host.
+type downwardAPIVolume struct {
+	volName                 string
+	fieldReferenceFileNames map[string]string
+	pod                     *api.Pod
+	plugin                  *downwardAPIPlugin
+	opts                    *volume.VolumeOptions
+	mounter                 mount.Interface
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+// The internal sync interval of kubelet will drive the refresh of data.
```

Comment:
I am worried about this longer term, but I guess it is OK for now.  long term it should have its own ticker an refresh loop.